### PR TITLE
sign_in via google and facebook

### PR DESCRIPTION
**Pull request by Thanhdong0909**
Screenshot from 2019-06-11 09-29-46

### CODE REVIEW

**Comment by Thanhdong0909** [comment]
ready

File: app/controllers/omniauth_callbacks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+  def google_oauth2
+    generic_callback( "google_oauth2" )
+  end
+
+  def generic_callback(provider)
+    @identity = User.from_omniauth(request.env["omniauth.auth"])
+
+    @user = @identity || current_user
+    if @user.persisted?
+      sign_in_and_redirect @user, :event => :authentication
```
Comment (VuThiTranVan) [inline app/controllers/omniauth_callbacks_controller.rb]: `:event =>` thành `event: `

File: app/models/user.rb [hunk 2]
```diff
@@ -9,6 +10,26 @@ class User < ApplicationRecord
     length: {maximum: Settings.length_name_maximum}
   scope :user_activated, ->{where activated: true}
 
+  def self.new_with_session params, session
+    super.tap do |user|
+      if data = session["devise.facebook_data"] &&
+        session["devise.facebook_data"]["extra"]["raw_info"]
+        user.email = data["email"] if user.email.blank?
```
Comment (VuThiTranVan) [inline app/models/user.rb]: `if user.email.blank?` thành ` unless user&.email` hì sao nhỉ

**Comment by Thanhdong0909** [comment]
ready

**Pull request merged**